zallet/src/components/keystore.rs

```diff
@@ -0,0 +1,568 @@
+        callbacks: C,
+    ) -> Result<Option<age::IdentityFile<age::NoCallbacks>>, Error> {
+        let encrypted_identities = match &self.encrypted_identities {
+            Some(data) => data,
+            // If the keystore isn't encrypted, we don't need to do anything.
+            None => return Ok(None),
+        };
+
+        let decryptor = age::Decryptor::new_buffered(age::armor::ArmoredReader::new(
+            encrypted_identities.as_slice(),
+        ))
+        .expect("validated on start");
+
+        let encrypted_identity = age::encrypted::EncryptedIdentity::new(decryptor, callbacks, None)
+            .expect("validated on start");
+
+        encrypted_identity
+            .decrypt(None)
+            .map(|identity_file| Some(identity_file.with_callbacks(age::NoCallbacks)))
+            .map_err(|e| ErrorKind::Generic.context(e).into())
+    }
+
+    /// Unlocks the keystore using the given passphrase.
+    ///
+    /// The keystore will be re-locked after `timeout` seconds. Calling this method again
+    /// before the existing timeout expires will reset the timeout.
+    pub(crate) async fn unlock(
+        &self,
+        passphrase: age::secrecy::SecretString,
+        timeout: u64,
+    ) -> bool {
+        // Prepare a callback that only responds to passphrase requests.
+        #[derive(Clone)]
+        struct PassphraseCallbacks(age::secrecy::SecretString);
+        impl age::Callbacks for PassphraseCallbacks {
+            fn display_message(&self, _: &str) {}
+            fn confirm(&self, _: &str, _: &str, _: Option<&str>) -> Option<bool> {
+                unreachable!()
+            }
+            fn request_public_string(&self, _: &str) -> Option<String> {
+                unreachable!()
+            }
+            fn request_passphrase(&self, _: &str) -> Option<age::secrecy::SecretString> {
+                Some(self.0.clone())
+            }
+        }
+
+        let identity_file = match self
+            .decrypt_identity_file(PassphraseCallbacks(passphrase))
+            .await
+        {
+            Ok(Some(identity_file)) => identity_file,
+            _ => return false,
+        };
+
+        let decrypted_identities = match identity_file.into_identities() {
+            Ok(identities) => identities,
+            Err(_) => return false,
+        };
+
+        // If there is an existing relock task, abort it so we don't race while writing
+        // the decrypted identities.
+        let mut relock_task = self.relock_task.lock().await;
+        if let Some((_, existing_timeout)) = relock_task.take() {
+            existing_timeout.abort();
+        }
+
+        *self.identities.write().await = decrypted_identities;
+
+        // Start a task to relock the keystore after the given timeout.
+        let duration = Duration::from_secs(timeout);
+        let identities = self.identities.clone();
+        *relock_task = Some((
+            SystemTime::now() + duration,
+            tokio::spawn(async move {
```

Comment:
This doesn't lock the mutex when it starts writing. So how do we know a write isn't in progress when we lock `self.relock_task` in order to change the timeout (in a concurrent call to `unlock`, say)?

Reply:
I don't believe there is a problem here; @daira does. I'm going to avoid the problem by awaiting `existing_timeout` after aborting it, which ensures that the write below can never be in play.